pkg/gardenlet/controller/shoot/shoot_control.go

```diff
@@ -786,6 +799,51 @@ func (c *Controller) isHibernationActive(ctx context.Context, shootSeedNamespace
 	return v1beta1helper.HibernationIsEnabled(shoot), nil
 }
 
+func needsControlPlaneDeployment(ctx context.Context, o *operation.Operation, kubeAPIServerDeploymentFound bool, infrastructure *extensionsv1alpha1.Infrastructure) (bool, error) {
+	var (
+		namespace = o.Shoot.SeedNamespace
+		name      = o.Shoot.GetInfo().Name
+	)
+
+	// If the `ControlPlane` resource and the kube-apiserver deployment do no longer exist then we don't want to re-deploy it.
+	// The reason for the second condition is that some providers inject a cloud-provider-config into the kube-apiserver deployment
+	// which is needed for it to run.
+	exists, markedForDeletion, err := extensionResourceStillExists(ctx, o.K8sSeedClient.APIReader(), &extensionsv1alpha1.ControlPlane{}, namespace, name)
+	if err != nil {
+		return false, err
+	}
+	// treat `ControlPlane` in deletion as if it is already gone. If it is marked for deletion, we also shouldn't wait
+	// for it to be reconciled, as it can potentially block the whole deletion flow (deletion depends on other control
+	// plane components like kcm and grm) which are scaled up later in the flow
+	if (!exists || markedForDeletion) && !kubeAPIServerDeploymentFound {
+		return false, nil
+	}
+
+	// The infrastructure resource has not been found, no need to redeploy the control plane
+	if infrastructure == nil {
+		return false, nil
+	}
+
+	if providerStatus := infrastructure.Status.ProviderStatus; providerStatus != nil {
+		// The infrastructure resource has been found with a non-nil provider status, so redeploy the control plane
+		return true, nil
+	}
+
+	// The infrastructure resource has been found, but its provider status is nil
+	// In this case the control plane could not have been created at all, so no need to redeploy it
+	return false, nil
```

Comment:
Nit: I think replacing this series of ifs with a `switch` would be slightly easier to read and understand.